File: python/n107_levelTravel.py
# https://leetcode-cn.com/problems/binary-tree-level-order-traversal-ii/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode):
        level_trav = []
        cur_queue = []
        if root == None:
            return level_trav
        cur_queue.append(root)      
        while cur_queue != []:
            next_queue = []
            cur_lev = []
            while cur_queue != []:
                cur_node = cur_queue.pop(0)                              
                cur_lev.append(cur_node.val)
                if cur_node.left is not None: next_queue.append(cur_node.left)
                if cur_node.right is not None: next_queue.append(cur_node.right)
            level_trav.append(cur_lev)
            cur_queue = next_queue
        return level_trav[::-1] 


if __name__ == "__main__":
    pass

File: cpp/n107_levelTravel.cpp
// https://leetcode-cn.com/problems/binary-tree-level-order-traversal-ii/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> level_trav = {};
        if (root == NULL) return level_trav;
        vector<TreeNode*> cur_queue = {};
        cur_queue.push_back(root);
        while(!cur_queue.empty()){
            vector<TreeNode*> next_queue = {};
            vector<int> cur_lev = {};
            while(!cur_queue.empty()){
                TreeNode* cur_node = cur_queue.front();
                cur_queue.erase(cur_queue.begin());
                cur_lev.push_back(cur_node -> val);
                if (cur_node -> left != NULL) next_queue.push_back(cur_node -> left);
                if (cur_node -> right != NULL) next_queue.push_back(cur_node -> right);
            }
            level_trav.push_back(cur_lev);
            cur_queue = next_queue;
        }
        reverse(level_trav.begin(), level_trav.end());
        return level_trav;
    }
};

int main(){
    return 0;
}